### ISSUE TITLE
Add a PPA or APT repo

### ISSUE THREAD

**Issue by pizzadude** [Eloston/ungoogled-chromium#37]
Hi, interesting project.

Is it possible for you to setup a Ubuntu PPA so Ubuntu/Mint/[insert derivative here] users could easily be up to date?

**Comment by Eloston** [comment]
I like the idea, but I don't have time to maintain a PPA manually. If this can be automated in some way, I would like to hear it.

If you or someone else would like to maintain a PPA or repository, go right ahead. I can add a link to it from the README.

**Comment by pizzadude** [comment]
I'm not sure how to make a PPA, I'm not a developer. But PPA's can be automated. http://askubuntu.com/questions/550012/how-to-create-a-automatic-ppa

Sorry if this is too much to ask. :<

**Comment by Eloston** [comment]
Well the link you provided basically says that a CI can push packages to a PPA server, which means that I would need to figure out how to use Travis CI (or some other CI system) and find a server to host a PPA.

That sounds nice, so I might look into this in the future when I have more time.

**Comment by hrj** [comment]
@Eloston As a first step, you could also consider pushing the build from Travis-CI to Github Releases pages as per [this doc](https://docs.travis-ci.com/user/deployment/releases/). That will take care of hosting atleast.

**Comment by Eloston** [comment]
@hrj: I'm not sure if I trust Travis CI enough to give it the [`public_repo` permission](https://developer.github.com/v3/oauth/#scopes) to this account. But what really bugs me is that they telll you to use their magic command `travis setup releases` that obscures how things are setup:

> Instead of setting it up manually, it is highly recommended to use `travis setup releases`, which will automatically create a GitHub oauth token with the correct scopes and encrypts it.

I could make another GitHub account that mirrors this repository and integrates with Travis CI, but that would require me to manage another GitHub account on top of Travis CI. Though this option seems more appealing the more I think about it.

**Comment by hrj** [comment]
@Eloston Oh, that's an overarching permission. I understand your hesitation.

I wonder how the permission system works. If you create a dummy Github account, add it as a collaborator on this repository (without write permissions), create an oauth token for the dummy account and give it to Travis, will Github allow code writes through that token? If it allows releases but not code-writes, then it might be worth exploring.

**Comment by Eloston** [comment]
@hrj: I'm pretty sure you have to own the repository. I'm a collaborator on another repository (with write access), and I couldn't get Travis CI to see it.

Fortunately, there are commands in git that make it easy to mirror a remote repository. [Here's GitHub's documentation on them](https://help.github.com/articles/duplicating-a-repository/). This way my dummy account can be separate from this account.

**Comment by hrj** [comment]
@Eloston Ah, ok.

FWIW, Github seems to have come out with a [new auth system](https://developer.github.com/changes/2016-09-14-Integrations-Early-Access/) yesterday. Here's hoping that Travis-CI leaps on to it quickly.

**Comment by Eloston** [comment]
@hrj: Huh, looks interesting. That would solve the hassle of setting up a separate account.

I'm a bit worried about the build environment they provide us. They allocate 7.5 GB (since I have to use a Docker image of Ubuntu Xenial), which might be enough to link if I set the right flags. However, I'm not sure how much disk space they allocate -- my whole build directory is taking 4 GB, which is pretty small by Chromium standards. Also, I'm not clear what Travis considers a "job". From what I've read, a single job can only last for about 50 minutes maximum (assuming it prints output at at least 10 minute intervals), but the build can take as long as it needs to. Are jobs the processes Travis directly invokes? Does this include subprocesses? Threads? A build takes about 3 hours, so the definition is pretty crucial.

**Comment by podshumok** [comment]
Launchpad can automate on-commit build, packaging and publishing of deb packages
You just need
1) a debian/ directory
2) bzr-script https://help.launchpad.net/Packaging/SourceBuilds/Recipes
3) setup launchpad mirror of source repo

**Comment by Eloston** [comment]
@podshumok Good to know. I might consider it if Travis doesn't work out. Would you mind linking some documentation to do this? Thanks.

**Comment by podshumok** [comment]
Some docs are available here https://help.launchpad.net/Packaging/SourceBuilds/GettingStarted
The hardest part is to actually make a debian/ directory, everything else is more or less obvious, although may seem not so easy at the beginning...

You may want to look at https://code.launchpad.net/~chromium-team/chromium-browser/chromium-browser.head

Best thing is that Launchpad provides separate everyday builds for different platforms and Ubuntu versions

**Comment by Eloston** [comment]
@podshumok Alright thanks.

**Comment by MikolajQ** [comment]
PPA would be great!
+1 for PPA

**Comment by craft37** [comment]
Please add the PPA. 
And great thanks for you work, really.

**Comment by Lesik** [comment]
I oppose the use of PPAs/Launchpad.

I don't understand why using the OBS as suggested in #17 and #59 was rejected. The OBS allows building binaries for Debian, Ubuntu, Fedora, CentOS/RHEL, Arch Linux and many more distributions. By investing some time into this you receive a unified build/hosting service serving packages for all distributions, not just Ubuntu. Launchpad, on the other hand, **only** supports Ubuntu (and while it would theoretically work to install a PPA on Debian, you should [**never**](https://wiki.debian.org/DontBreakDebian#Don.27t_make_a_FrankenDebian) do [that](https://unix.stackexchange.com/questions/83060)!)

**Comment by hrj** [comment]
> I don't understand why using the OBS as suggested in #17 and #59 was rejected.

It wasn't rejected. #59 was closed as a duplicate of #17, which is still open.

**Comment by zabbal** [comment]
Using OBS does not contradict using Launchpad - those are orthogonal things. Having ppa would be really convenient for Ubuntu users which would greatly help with wider adoption and testing.

**Comment by Lesik** [comment]
@zabbal How do Ubuntu users benefit from Launchpad more than from OBS? Why maintain two services? OBS does everything Launchpad can, and much more. Or am I mistaken?

**Comment by message** [comment]
bump

**Comment by Eloston** [comment]
I'm not quite sure how APT or PPA repos work, but are they basically retrieving files from a web server? If so, it may be possible to generate these files and upload them to the GitHub pages repo of the downloads repository.

**Comment by pseudoj** [comment]
+1 for a PPA repository.

**Comment by craft37** [comment]
@Eloston Personal Package Archive (PPA) only works with launchpad
https://askubuntu.com/questions/71510/how-do-i-create-a-ppa

**Comment by Eloston** [comment]
@craft37 Thanks for that link. [This is what I was curious about earlier](http://askubuntu.com/a/493577), and it seems it is possible. We can work on integrating this approach to the new WIP downloads system in the future.

It is unfortunate that PPAs are tied to Launchpad, but there's no significant benefit of also having a PPA.

**Comment by message** [comment]
> but there's no significant benefit of also having a PPA.

Well, i can get updates when they are ready, and not checking every month for fresh release.

**Comment by Eloston** [comment]
@message 
> but there's no significant benefit of **also** having a PPA.

Bolded for emphasis.

**Comment by mtimofiiv** [comment]
I might be late to the party, but...

> @hrj: I'm pretty sure you have to own the repository. I'm a collaborator on another repository (with write access), and I couldn't get Travis CI to see it.

@Eloston if you make an organisation on GH and move the repo into it, all users that are part of the org should be able to add via Travis.

**Comment by Eloston** [comment]
@mtimofiiv Travis is not designed to compile Chromium. No free system out there that I know of will like people using their services to build Chromium.

**Comment by hrj** [comment]
@Eloston Probably, the [GitLab CI](https://about.gitlab.com/gitlab-ci/) system would suit this project. I haven't used it, but from what I have read:
* Works similar to Travis-CI
* It's open-source.
* Can use our own build machines or use the ones provided on gitlab.com

---

Aside, is the cost of build machines a constraint? If so, I can chip in. I guess a single $5 per month Digital Ocean instance would be sufficient (?). And if more people contribute, one could run multiple builds in parallel (either for different platforms, or for cross-verification).

**Comment by fedenko** [comment]
+1 for a PPA

**Comment by sixtyfive** [comment]
@Eloston Is there any way to get in touch with you privately? GPG mail, etc.?

**Comment by Eloston** [comment]
@sixtyfive If it's related to my open-source projects, I prefer public communication. Otherwise, you can direct message me on Gitter (there is a link to Gitter from the README).

**Comment by olosz** [comment]
+1 for a PPA.
Keep up the good work.

**Comment by dannycolin** [comment]
PPA is too Ubuntu-centric. I'll prefer to see ungoogled-chromium directly in the Debian Repository. That way, it can easily be push to the ubuntu repository as well (an any other Debian/Ubuntu-based distro).

**Comment by Eloston** [comment]
One hacky method to accomplish this is to setup a flat APT repo in GitHub Releases of the https://github.com/ungoogled-software/ungoogled-chromium-binaries/ repository. Each distribution and version combination (e.g. Debian Stretch) will have its own release (e.g. `debian_stretch`). Then, a script can be used to update any GitHub release with new APT repository info.

**Comment by T-vK** [comment]
Instead of running Gitlab on a rented server, how about using gitlab.com directly? They allow up to 2000 build minutes (=33 hours) per month. And in case we should ever need more, we could upgrade for $19 to get 10000 minutes. I'd be more than willing to donate. But for now 33 hours should be plenty.  

I have used gitlab-ci before. If you're interested I could try to get it to build ungoogled-chromium and if I succeed I could give you the instructions.

**Comment by Eloston** [comment]
@T-vK I wouldn't mind this for the upcoming `linux_portable` config type; you could build it once then package using `linux_simple` (for a `.tar.xz`) and Debian `minimal` (for a `.deb`).

Right now, there are some rough edges around the build and packaging processes for those types I listed, so I'd hold off if your goal is to make something that can build releases right now. I'm going to be making some changes which may break some scripts you would write.

However, it wouldn't be a bad idea to see if GitLab CI can actually build ungoogled-chromium from a computational resources point of view. If you're curious, you can try it out and let us know how it goes.

**Comment by T-vK** [comment]
I was too curious too see what would happen and already started it a few hours ago. It seems to work so far (lots of warnings, no errors). I think the build process will take around 5 hours actually.  
I just found a limitation though. It will only display the first 4Mb of the console output.  

You can take a look at the build process here: https://gitlab.com/T-vK/ungoogled-chromium/-/jobs/47589261

Is there a way to make the build output less verbose? Maybe don't display warnings, only errors?

**Comment by Eloston** [comment]
@T-vK Oh, neat. Kinda sucks the log got truncated, though.

> Is there a way to make the build output less verbose? Maybe don't display warnings, only errors?

The only way I know of is to patch out the unimplemented warning options from GN files. I'm not really sure if there's a way to exclude warning output altogether.

**Comment by T-vK** [comment]
Well, I could pipe the output to grep and strip out all lines containing the word "warning". But if an error, for some reason, contains the word "warning", then we won't see it.  
I could also use tee to pipe the unfiltered output to a file. But it would be a lot of work to keep the script from exiting after the build failed in order to send the file to some place where we could read it.

**Comment by hrj** [comment]
Nice to see progress. FYI, the Gitlab build configuration allows you to specify build artifacts. These are available for download for every pipeline/job. You could mark the `tee` output as a build artifact and thus not worry about the 4MB limit. You could also use the same idea to mark the generated `.deb` as a build artifact allowing anyone to download it.

(I am familiar with this process but currently travelling; so I won't have time for another week to contribute directly)

**Comment by Eloston** [comment]
Nice @T-vK. Is it possible to trigger the build when a new tag is pushed?

Also, is there some way for this to integrate with a PPA or APT repo? If not, we should move this discussion elsewhere.

**Comment by T-vK** [comment]
Just found another limitation: The artifacts can't be bigger than 100Mb: https://gitlab.com/T-vK/ungoogled-chromium/-/jobs/47700776

I probably shouldn't have specified the whole build directory...  
I just started another build and this time I tried to only mark the `*.deb`, `*..changes`, and `*.buildinfo` files as artifacts. These are only a little over 56Mb at the moment. If this becomes a problem in the future we can still do a simple `curl` or `ftp` upload to some other place. 

Here is the new build: https://gitlab.com/T-vK/ungoogled-chromium/-/jobs/47772535

@Eloston 
I think it is possible to have a new tag be the trigger. About the PPA thingy. I would say, if it can be scripted in bash, then it can be done. Unless there is another unexpected limitation.

**Comment by Eloston** [comment]
@T-vK Thanks for all the work. Due to the build artifact restriction, I think it would be better to just build a `.tar.xz` of `linux_portable` because there's not enough capacity to store different packages.

Debian `source packages` are pretty straight-forward. Basically, it consists of a `.dsc`, a tar of the original source code, and a tar of a `debian` directory (or a compressed patch file that adds the `debian` directory). They're the same files you get when you want to build any Debian package. I was thinking about adding a packaging method for it to fix #20 at some point.

**Comment by T-vK** [comment]
@Eloston 
From what I understand, the artifact size limit is per job. So we could simply add more build jobs and build them all in parallel. But if you actuallly add the option to create a source package, we could just upload it to launchpad and let them do the hosting for the .deb packages.

**Comment by Eloston** [comment]
@T-vK To clarify, I mean building different types of packages, such as `.deb`, `.tar.xz`, or even AppImages and Flatpacks from the same build output. But since a single packaging type is about 50 MB, that isn't feasible.

**Comment by T-vK** [comment]
@Eloston 
But does it really have to be the same output? I mean we could for instance first do some compiling in the before_script and then have multiple packaging jobs that create different packages. 
So we could have 100Mb for the debian packages and 100Mb for the linux_portable thing etc.

I tried to build the linux_portable btw and it failed: https://gitlab.com/T-vK/ungoogled-chromium-mirror/-/jobs/47846500
`Not an existing flavor: 'minimal'`   
Did I misunderstand the build instructions? I used `./utilikit/generate_build_files.py debian --flavor minimal --apply-domain-substitution`

**Comment by Eloston** [comment]
@T-vK 

> we could for instance first do some compiling in the before_script and then have multiple packaging jobs that create different packages.

I wasn't aware that was possible. If all the compiling of `linux_portable` could be done in `before_script`, then that would save a lot of hassle of packaging that same binary in multiple different formats.

> I tried to build the linux_portable btw and it failed

You mean you failed to generate the packaging scripts for Debian `minimal`? [That's expected](https://github.com/Eloston/ungoogled-chromium/wiki/statuses#packaging-types). To clarify, `linux_portable` is a configuration type (soon to be renamed to base config bundle; more on that after I finish writing `buildkit` along with the documentation), which is separate from packaging types.

You should be able to build `linux_portable` base bundle with `linux_simple` packaging type with 1ff74286a677ed865b2dd46d1eec98cb8a46b077. [Preliminary instructions are here](https://github.com/Eloston/ungoogled-chromium/blob/1ff74286a677ed865b2dd46d1eec98cb8a46b077/BUILDING.md#other-linux-distributions); note the typo that `UTILIKIT_CONFIG_TYPE=linux_simple` should be `linux_portable`.

**Comment by T-vK** [comment]
I was trying to compile the linux_portable, but I guess I misunderstood the building instructions and to be honest I still don't really understand them. There are lots of things that aren't clear to me.

At the moment the before_script just looks like this:
before_script:

``` bash
apt-get -qq update --yes
apt-get -qq install --yes packaging-dev python3 python ninja-build > /dev/null
mkdir build/
mkdir build/sandbox
mkdir build/downloads
```

and the debian package job looks like this:
``` bash
./utilikit/prepare_sources.py
./utilikit/substitute_domains.py
./utilikit/generate_build_files.py debian --flavor standard --apply-domain-substitution
cd build/sandbox
# dpkg-checkbuilddeps # Checks and reports any additional packages needed
echo -en "y\n" | mk-build-deps -i debian/control > /dev/null # install any additional packages needed
dpkg-buildpackage -b -uc | grep -v "warning"
```

and that worked fine.

Now the question is what should I move into the before_script and how would the debian package job change? 

From the new instructions you linked I would assume something like this:

before_script:

``` bash
apt-get -qq update --yes
apt-get -qq install --yes packaging-dev python3 python ninja-build > /dev/null
export UTILIKIT_CONFIG_TYPE=linux_portable
./utilikit/check_requirements.py
mkdir build/
mkdir build/sandbox
mkdir build/downloads
./utilikit/prepare_sources.py
./utilikit/substitute_domains.py
# ./utilikit/generate_build_files.py # I assume this would't work for linux_portable?
# ./utilikit/export_resources.py # Is this required in order to get the patches?
# "2. Apply patches" (Is "1." missing?) I've never used .patch files, how do you apply them?
# "Build GN via tools/gn/bootstrap/bootstrap.py" - There is no such path in the project. Will this be generated? Or did you mean `./utilikit/build_gn.py` ? 
#  "Run gn gen with the GN flags" - Do I have to parse `resources/configs/linux_portable/gn_flags
` and add these lines as parameters to the command `gn gen`?
# "Build chromium via ninja" - now I'm really lost.
# "Package the build outputs" # using ./utilikit/archive_packager.py ? Where is `FILES.cfg for `--files-cfg `? 
```

For the debian build job script I can only guess.

**Comment by Eloston** [comment]
Are you reading `General building instructions` under `Building generalizations and additional information`? You don't need to read that; I linked the `Other Linux distributions` section which is above all that. The section ends at the line "where `OUTPUT_FILE` is the path for the new archive". It's understandable if you find the documentation confusing; I'm rewriting it along with `buildkit`.

I took a look at the GitLab CI documentation, and I think I'm starting to understand what you mean by "before_script" and "jobs". So here's a rough pipeline for building and packaging `linux_portable`:

1. Install all build dependencies in `before_script` or have it installed in a pre-built Docker image (as you have suggested).
2. In the Build stage, there is one job that downloads and builds the code, and packages the necessary files into a `.tar.xz`. The `.tar.xz` will be the build artifact for this job.
3. In a Packaging stage, there can be multiple simultaneous jobs that repackage the xz-compressed tar archive into different formats; e.g. AppImage, Flatpak, Debian packages
    * Note that both `standard` and `minimal` flavors build and package with the same command. I think there may be a way to tell `dpkg-buildpackage` to package only

Then all of the packages, including the `.tar.xz`, can be artifacts available for download.

**Comment by T-vK** [comment]
Okay. For step 2, what are the "necessary files"? The whole `build` directory? If so, then we might not get away with passing that to another stage because of the 100Mb limit, but I'm not sure.  
Basically if I understood you correctly, it would have to look somewhat like this:

``` yml
image: debian:stretch

before_script:
  - apt-get -qq update --yes
  - apt-get -qq install --yes packaging-dev python3 python ninja-build > /dev/null
  - mkdir -p build/{downloads,sandbox} 

stages:
  - build
  - package

build:
  stage: build
  only:
    - tags
    - triggers
  script:
    - export UTILIKIT_CONFIG_TYPE=linux_simple
    - ./utilikit/prepare_sources.py
    - ./utilikit/substitute_domains.py
  artifacts:
    paths:
      - ./build/

debian_package:
  stage: package
  only:
    - tags
    - triggers
  dependencies:
    - build
  script:
#   - ./utilikit/generate_build_files.py debian --flavor minimal --apply-domain-substitution
    - ./utilikit/generate_build_files.py debian --flavor standard --apply-domain-substitution
    - patch < resources/packaging/debian/minimal.patch
    - cd build/sandbox
    - echo -en "y\n" | mk-build-deps -i debian/control > /dev/null # install any additional packages needed
#   - # If necessary, change the dependencies in debian/control and modify CLANG_BASE_PATH in debian/rules to accomodate your environment.
    - dpkg-buildpackage -b -uc
    - cd ../../
  artifacts:
    paths:
    - ./build/*.*

archive:
  stage: package
  only:
    - tags
    - triggers
  dependencies:
    - build
  script:
    - ./utilikit/generate_build_files.py linux_simple --apply-domain-substitution
    - cd build/sandbox
#   - # Use "export CLANG_BASE_PATH=/path/to/clang_files" if Clang and related files are not located under /usr
    - ./ungoogled_linux_simple/build.sh
    - cd ../../
    - ./utilikit/archive_packager.py --files-cfg chrome/tools/build/linux/FILES.cfg --archive-format tar_xz --build-output-dir out/Default --target-cpu auto --output-file ./build/ungoogled-chromium.tar.xz
  artifacts:
    paths:
    - ./build/ungoogled-chromium.tar.xz
```

**Comment by Eloston** [comment]
> Okay. For step 2, what are the "necessary files"?

I should've been a bit more clear: only the necessary files *for running Chromium* need to be passed from the Build stage to the Packaging stage. The goal of the build stage is to download all the code necessary to build ungoogled-chromium, and produce the final binaries and supporting files. The goal of the packaging stage is to produce different formats of said binaries and supporting resources to give users options.

To be more specific, here are the steps that can be performed at various stages *(these are untested and may have mistakes)*:

Here's a hack to ensure all the proper requirements are installed in `before_script`:

```
export UTILIKIT_CONFIG_TYPE=linux_simple
# Converting the standard flavor into minimal because it saves the hassle later.
pushd resources/packaging/debian/standard
patch -p1 < minimal.patch
popd
tmpdebdir=/tmp/tmpdebdir
mkdir $tmpdebdir
./utilikit/generate_build_files.py --output-dir $tmpdebdir debian --flavor standard --apply-domain-substitution
printf "y\n" | mk-build-deps -i $tmpdebdir/debian/control > /dev/null
rm -rf $tmpdebdir
```

Here's what can be done in the Build stage:

```
export UTILIKIT_CONFIG_TYPE=linux_simple
./utilikit/prepare_sources.py
./utilikit/substitute_domains.py
./utilikit/generate_build_files.py linux_simple --apply-domain-substitution
cd build/sandbox
# Use "export CLANG_BASE_PATH=/path/to/clang_files" if Clang and related files are not located under /usr
./ungoogled_linux_simple/build.sh
../../utilikit/archive_packager.py --files-cfg chrome/tools/build/linux/FILES.cfg --archive-format tar_xz --build-output-dir out/Default --target-cpu auto --output-file $OUTPUT_FILE
```
where `$OUTPUT_FILE` is the build artifact for the only job of the Build stage.

I don't know how to create Debian packages from the `.tar.xz`, but maybe something like this could work?:

```
export UTILIKIT_CONFIG_TYPE=linux_simple
sandboxdir=/tmp/debpkgs/sandbox
# I'm assuming the conversion from the standard to minimal flavors was preserved from before_script
./utilikit/generate_build_files.py --output-dir $sandboxdir debian --flavor standard --apply-domain-substitution
mkdir -p $sandboxdir
pushd $sandboxdir
mkdir -p out/Default
pushd out/Default
tar xf /path/to/build_artifact.tar.xz
popd
dpkg-buildpackage -b -uc -nc
```

and `.deb` packages would appear in `/tmp/debpks`, maybe. You might need to do some experimentation with `dpkg-buildpackage` if that doesn't produce binaries.

> I tried it on the latest develop branch and it fails. Any ideas? https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/47967733

It seems that Python is having some troubles identifying the correct encoding for a file. I'll have to look more into it later. If I don't get `buildkit` done soon enough, I might push a fix for `utilikit` in the meantime.

**Comment by T-vK** [comment]
Okay I updated the [.gitlab-ci.yml](https://gitlab.com/T-vK/ungoogled-chromium-dev-test/blob/5d17aaec7b0661480ce77223d7179150aca969d4/.gitlab-ci.yml). And I also created a [Docker image](https://hub.docker.com/r/tavk/ungoogled-chromium-build-environment/). 
The build job fails after these two lines:
```
pushd resources/packaging/debian/standard
patch -p1 < ../minimal.patch
```
https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/48009184

**Comment by Eloston** [comment]
Hm, it seems the patch will need to be updated. Not sure why it failed in the first place.

**Comment by T-vK** [comment]
Too bad.. Well, let me know when it's fixed. I'd really like to see this working.

**Comment by dubvulture** [comment]
`patch -p1 --ignore-whitespace < ../minimal.patch` should temporarily solve this issue.

**Comment by T-vK** [comment]
Okay, I applied the PR and get a new error now: https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/48376514

**Comment by Eloston** [comment]
@T-vK  I shall bring your attention to https://github.com/Eloston/ungoogled-chromium/issues/37#issuecomment-357693096:

> > I tried it on the latest develop branch and it fails. Any ideas? https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/47967733
>
> It seems that Python is having some troubles identifying the correct encoding for a file. I'll have to look more into it later. If I don't get `buildkit` done soon enough, I might push a fix for `utilikit` in the meantime.

**Comment by T-vK** [comment]
@Eloston Any news/progress?

**Comment by Eloston** [comment]
@T-vK I got caught up with some stuff that pushed back work on ungoogled-chromium. I have finished designing buildkit, so I am currently implementing it.

Also, I have decided that all patches should be encoded in UTF-8. I'm guessing that all patches are already in UTF-8, so you can try1f8f80e3221fc09a5dbe903e3c2f57454d316d4b and see if anything breaks.

**Comment by T-vK** [comment]
@Eloston Nice. :) I applied the change and the error is gone now.
It seems like the download that `./utilikit/prepare_sources.py` tries to do is failing:  
https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/51663632

**Comment by Eloston** [comment]
Is there no `series` file in `ungoogled_linux_simple/patches`?

**Comment by T-vK** [comment]
https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/51767956 We shall see...

**Comment by Eloston** [comment]
`export UTILIKIT_CONFIG_TYPE=linux_simple`

`linux_simple` is not a valid config type. Do you mean `linux_portable`?

**Comment by T-vK** [comment]
Ah, sorry yes of course.. https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/51773290

/bin/sh: 1: ../../../../../../../bin/clang++: not found
ninja: build stopped: subcommand failed.

I assume I need to set CLANG_BASE_PATH? But I'm not sure to which directory.

**Comment by Eloston** [comment]
Depends on your system. Debian places LLVM 3.9 files under `/usr/lib/llvm-3.9`

**Comment by T-vK** [comment]
Seems to work. I guess this will take a few hours now: https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/51808431

**Comment by T-vK** [comment]
It seems to work now. Only the artifact uploading failed for some reason. I'll try to fox that.

**Comment by T-vK** [comment]
Success!  
Build: https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/52117512  
Artifacts: https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/52117512/artifacts/browse/build/

**Comment by T-vK** [comment]
@Eloston I have not been able to get the packaging job to work yet: https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/52322641

`dpkg-buildpackage` output:
```
$ dpkg-buildpackage -b -uc -nc
dpkg-buildpackage: info: source package ungoogled-chromium-browser
dpkg-buildpackage: info: source version 63.0.3239.132-1
dpkg-buildpackage: info: source distribution stable
dpkg-buildpackage: info: source changed by Maintainer <maintainer@null>
 dpkg-source --before-build sandbox
dpkg-buildpackage: info: host architecture amd64
 debian/rules build
 debian/rules binary
make: 'build' is up to date.
dh binary --parallel
   dh_testdir -O--parallel
   dh_update_autotools_config -O--parallel
   debian/rules override_dh_auto_configure
make[1]: Entering directory '/tmp/debpkgs/sandbox'
# output compiler information
g++ --version
g++ (Debian 6.3.0-18) 6.3.0 20170516
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

make[1]: Leaving directory '/tmp/debpkgs/sandbox'
   debian/rules override_dh_auto_build-arch
make[1]: Entering directory '/tmp/debpkgs/sandbox'
mkdir -p out/Default || true
./tools/gn/bootstrap/bootstrap.py -o out/Default/gn -s -j2
make[1]: ./tools/gn/bootstrap/bootstrap.py: Command not found
debian/rules:88: recipe for target 'out/Default/gn' failed
make[1]: *** [out/Default/gn] Error 127
make[1]: Leaving directory '/tmp/debpkgs/sandbox'
debian/rules:85: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: debian/rules binary gave error exit status 2
ERROR: Job failed: exit code 1
```

The job:
```
    - export UTILIKIT_CONFIG_TYPE=linux_portable
    - export sandboxdir=/tmp/debpkgs/sandbox
    - mkdir -p $sandboxdir
# I'm assuming the conversion from the standard to minimal flavors was preserved from before_script
    - ./utilikit/generate_build_files.py --output-dir $sandboxdir debian --flavor standard --apply-domain-substitution
    - pushd $sandboxdir
    - mkdir -p out/Default
    - pushd out/Default
    - wget https://gitlab.com/T-vK/ungoogled-chromium-dev-test/-/jobs/52117512/artifacts/raw/build/ungoogled-chromium.tar.xz
    - tar xf ungoogled-chromium.tar.xz
    - rm ungoogled-chromium.tar.xz
    - popd
    - dpkg-buildpackage -b -uc -nc
```

Any ideas?

**Comment by Eloston** [comment]
Hmm, there are two problems that'll need to be sorted out:

1. We need a command that invokes the "install" step of debhelper and beyond to produce packages, not try to attempt the "build" step.
2. Some files are being copied/moved around in the "build" step before packages are created (presumably in the "install" step), so these commands will need to be moved out of the "build" step to before packaging files are collected but during the "install" step.

I think I have an idea for Problem 2, but I don't know enough about the Debian packaging system to attempt Problem 1.

**Comment by dannycolin** [comment]
> @Eloston I think I have an idea for Problem 2, but I don't know enough about the Debian packaging system to attempt Problem 1.

You can find informations on that wiki page. There's some links pointing to more advance stuff too.
https://wiki.debian.org/Packaging

**Comment by Eloston** [comment]
@dannycolin Thanks, but the issue isn't information, it's time. If someone already knows how to do this, it'll save some time.

**Comment by Eloston** [comment]
If someone here wants to get a PPA for Ubuntu, it may be possible to build a source package with `dpkg-source -b tree` (if running within buildspace) and then upload the generated files to Launchpad to build. The code in `develop` should be working right now.

**Comment by T-vK** [comment]
@Eloston What exactly do you mean by `buildspace`? Do you mean the "build" directory after all the [commands from the before_script and build stage](https://gitlab.com/T-vK/ungoogled-chromium-dev-test/blob/f89edd43cf398cadb3c88554fce1861b6577e12f/.gitlab-ci.yml) have been executed?

**Comment by Eloston** [comment]
@T-vK Please read DESIGN.md in `develop`. If it's unclear, we can discuss that in a new issue.

**Comment by T-vK** [comment]
@Eloston `./buildkit-launcher.py getsrc` says: 

    2018-02-21 12:37:35,414 - ERROR: Buildspace downloads does not exist: buildspace/downloads

which is weird because I ran `mkdir -p buildspace/downloads`.

https://gitlab.com/T-vK/ungoogled-chromium-clone-2018-02-20/-/jobs/53630707

**Comment by Eloston** [comment]
@T-vK Please try f031462978e20edff44079f7c5313357d22fd759 and see if you can get more information.

**Comment by T-vK** [comment]
Now I get:

    2018-02-22 09:54:37,556 - ERROR: Directory of file not found: [Errno 2] No such file or directory: '/builds/T-vK/ungoogled-chromium-clone-2018-02-20/buildspace/tree'

https://gitlab.com/T-vK/ungoogled-chromium-clone-2018-02-20/-/jobs/53815547

**Comment by Eloston** [comment]
@T-vK I fixed the buildspace tree problem in 4e3a14bad67a51bf2631513edc4fc1eb4c6b6bbf

Based on your error, it seems that `dpkg-source -b` doesn't work. Anyone have ideas?

**Comment by T-vK** [comment]
Since the official [Chromium is available on Launchpad](https://launchpad.net/chromium-browser), I wonder if there is a way to get the information on how this has been done and if it could be adjusted to work with ungoogled-chromium.

**Comment by Eloston** [comment]
It seems that the debian folder format I'm using (i.e. `3.0 (quilt)`) requires a orig tar file to **already be present**: 

```
   Format: 3.0 (quilt)
       Supported since dpkg 1.14.17.  A source package in this format contains
       at least an original tarball (.orig.tar.ext where ext can be  gz,  bz2,
       lzma  and  xz)  and  a  debian  tarball  (.debian.tar.ext).
```

We could add a custom target to `debian/rules` to accomplish this I believe (similar to what the official Debian `chromium` package does) but we would use the contents of `buildspace/tree` excluding the `debian` directory.

**Comment by T-vK** [comment]
Bad news, I got an email from GitLab:

> Dear T-vK,
> 
> At GitLab we are working hard to make GitLab.com as stable and reliable as it can be. But sometimes it means that we need to enforce some limits to make the service stable.
> 
> Since the beginning of CI at GitLab.com, our Shared Runners are available for all projects and are free of charge. However, to improve the reliability and general availability of Shared Runner for all GitLab.com users, we are forced to set a 3 hours build timeout for each job that is handled by these.
> 
> We have done a research to find out what is the best value for such timeout and we are convinced that 3 hours is a reasonable value, that will allow us to make the service stable and still affect as less projects as it can. At this moment, looking on the Project timeout setting in our database, we can say that enforcing chosen timeout may affect 584 projects. This means only 0.013% of all projects on GitLab.com.
> 
> We wrote ‘may be affected’, because not every job in projects that are using a bigger timeouts is indeed running for so long. But starting on May 8, 2018, a 3 hour timeout set on Shared Runners will take a precedence over the timeout configured in Project settings.
> 
> You are receiving this email because some of your GitLab.com projects are using Shared Runners and are specifying a bigger timeout than we plan to enforce:
> 
>     ungoogled-chromium-dev-test uses timeout: 1mo 4d 17h 20m
>     ungoogled-chromium uses timeout: 1mo 4d 17h 20m
>     Starting on May 8, 2018, timeout setting in specified projects will be overwritten by Shared Runners setting.
> 
>     If any of your jobs need to work for a longer time than 3 hours, you may still provide your own Runners that will handle them. Please read the documentation at https://docs.gitlab.com/ce/ci/runners/README.html if you need a guidance for how to do this.
> 
>     If you have any questions, please contact support@gitlab.com.
> 
>     Sincerely,
>     GitLab Support Team

**Comment by Eloston** [comment]
@T-vK That's not really a huge problem. If PPAs are the main interest for this issue, then the refactoring of Debian packaging after #367 will make it possible to offload the compilation to Launchpad.

**Comment by Eloston** [comment]
[It's finally here](https://github.com/Eloston/ungoogled-chromium/blob/master/docs/building.md#building-via-source-package). I haven't tested this yet, so I don't know what will happen.

**Comment by Eloston** [comment]
I have fixed source package generation as of 9178feeb6f075611d312d3792882fbc0abc33636. Extraction via `dpkg-source -x` also works.

I don't have the time to build right now. Anyone want to try building? Maybe uploading them to Launchpad?

**Comment by T-vK** [comment]
I'll try. https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92052639

**Comment by Eloston** [comment]
> I'll try. https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92052639
> Looks like the build worked: https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92052639/artifacts/browse/build/

I don't seem to have permission to view them?

> First of all Launchpad seems to want a PackageName_Version_source.changes file which doesn't seem to get generated during the build process.

I did some research, and it seems that .changes files can be generated by either binary or source package generation.

[According to Ask Ubuntu](https://askubuntu.com/questions/155410/how-do-i-add-create-a-source-changes-file), it seems we want to replace

```sh
cd ..
dpkg-source -b -uc
```

with 

```sh
debuild -S -sa
```

You'll also need to setup a GPG key so that the `debuild` command can sign the files. I don't know how to do this yet. Another thing to note is that you may need to modify `debian/changelog` so that a corresponding GPG key is chosen for signing (which I'm guessing based on the above Ask Ubuntu link).

> And secondly they are talking about an ubuntu suite in the documentation without explaining what the hell that is supposed mean.

In Debian, [suites and codenames are separate concepts](https://wiki.debian.org/DebianRepository/Format#Suite). In Ubuntu, [suites seem to be codenames according to packages.ubuntu.com](https://packages.ubuntu.com/search?suite=artful&section=all&arch=any&keywords=nonexistantname&searchon=names) (notice the top of the webpage: `Search in specific suite: [trusty] [trusty-updates] [trusty-backports] [xenial] [xenial-updates] [xenial-backports] [artful] [artful-updates] [artful-backports] [bionic] [bionic-updates] [bionic-backports] [cosmic]`).

**Comment by T-vK** [comment]
Sorry, I forgot to set the repo to be public. The links should work now.

> 
> According to Ask Ubuntu, it seems we want to replace
> cd ..
> dpkg-source -b -uc

In the building.md under `Building via source package` you say `dpkg-source -b src`? Is that what I should replace with `debuild -S -sa`?

I already have a GPG key that I generated some months ago and added to my launchpad account. I guess I'll have to dig out the private key again if we need it outside of launchpad.

Thanks for digging out the information on the suit thingy. I was not able to find that. I guess I'll just put in `stretch` then if I want to build for Debian Stretch?

**Comment by Eloston** [comment]
> In the building.md under Building via source package you say dpkg-source -b src. Is that what I should replace with debuild -S -sa?

Sorry, I made a small typo in my comment; `dpkg-source -b -uc` is supposed to be `dpkg-source -b src`. But, the intent remains the same; the commands `cd ..` *and* `dpkg-source -b src` should be replaced with `debuild -S -sa`.

> I already have a GPG key that I generated some months ago and added to my launchpad account. I guess I'll have to dig out the private key again if we need it outside of launchpad.

:+1: 

> Thanks for digging out the information on the suit thingy. I was not able to find that. I guess I'll just put in stretch then if I want to build for Debian Stretch?

[According to the Launchpad docs you sent me](https://help.launchpad.net/Packaging/PPA/Uploading#Using_packages_from_other_distributions):  `You can upload a source from any Debian-compatible distribution straight to your PPA with no changes required and it will be built and published in the targeted Ubuntu suite.`

If you want Debian packages, you might have to send it to something like the [openSUSE Build Service](https://build.opensuse.org/), since they seem to [provide an APT repo](https://en.opensuse.org/openSUSE:Build_Service_Debian_builds#Configuring_sources.list).

If we are okay with abusing git and/or potentially using Git LFS, we can do something like the following:
* Create another repo under ungoogled-software like `ungoogled-chromium-apt` that will hold only one commit with the binary and metadata files needed by APT.
* I can create another account that has write access only to ungoogled-chromium-apt that is accessible by a CI system (CI systems have a way to hide credentials from the public)
* Trigger the CI system via changes in ungoogled-chromium-binaries, and activate a script I can write that detects any changes to binaries for Debian-based systems and push only what has changed to ungoogled-chromium-apt.

Granted this will take a little bit of time, whereas OBS already solves this APT repo hosting and has the added benefit of building the packages for us.

**Comment by T-vK** [comment]
I see. I'll try to get it to work with ubuntu bionic then. If we get the launchpad thing to work, I might also try to get it to work with OBS. I'm not sure if I understand the git abusing idea correctly. Where would the binaries come from? I mean we can't build them in CI because of the time limitations, right? I think it would be much nicer to have a fully automated solution.

I found the GPG private key btw. But importing it to the CI docker container turns out to be very difficult.  I'll keep trying tough.

**Comment by Eloston** [comment]
@T-vK I made a small tweak to Ubuntu bionic packaging that affects the resulting package name: 0e1461ec690a0559fc8a045372423205356a1ad5

**Comment by T-vK** [comment]
Maybe this will fix this issue? https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92514801

I pulled the changes and started another build: https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92519818

**Comment by Eloston** [comment]
@T-vK 

> dpkg-source: error: cannot represent change to ungoogled-chromium-browser-build-deps_68.0.3440.106-1~buster_all.deb: binary file contents changed

It looks like you used `mk-build-deps` to install the build dependencies, but forgot to remove the resulting `.deb` package.

**Comment by T-vK** [comment]
@Eloston Thanks. I just did another build: https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92569956
Looks like I have some more work to do to make it recognize the imported GPG key..

**Comment by Eloston** [comment]
> I'm getting a lot of these:

Not entirely sure, but a lot of it is due to minified JS causing the column length check to go crazy. Are those files actually missing though?

But this one is interesting: `E: ungoogled-chromium-browser source: uploader-address-malformed Maintainer <maintainer@null>`

This may be causing a problem for the second problem you described with `No such file or directory` if a file fails to be generated as a result. Hopefully I fixed it in 23ebfe4ba694e61fc179edae79dcb7685d951e6d

On a tangential note, I'm surprised at how deep the checks go; they reported spelling errors in patch descriptions and an old URLs in debian/control.

**Comment by T-vK** [comment]
I pulled the changes and started another build. https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/92885375

I am replacing `-- ungoogled-chromium Authors <maintainer@null>` from the `packaging/ubuntu_bionic/changelog.ungoogin` file before the build btw because it seemed like the data has to match with the user ID of the GPG key.

I haven't had the time to check if the files are actually missing.

**Comment by Eloston** [comment]
@T-vK Sorry for the delayed response. Are you able to get more details on that error? The CI job doesn't show where in the make target `get-orig-source` failed.

**Comment by message** [comment]
This ticket was created 2 years ago today. Congratulations :tada:

**Comment by Eloston** [comment]
@message Thanks! Do you have something you want to contribute?

**Comment by T-vK** [comment]
@Eloston 
I think that error was caused by me trying to replace the `` thing with the GPG user ID. I undid that change and ran the job again. This time I'm getting `make: *** [get-orig-source] Error 255`:
https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/94960498

I'm not sure what that means. It's really hard for me to debug because I don't have Debian or Ubuntu. So I have to constantly make changes, commit and push them, wait for Gitlab CI to build it and then check the build output for errors. 
The error is coming form the `debian/rules get-orig-source`.

**Comment by Eloston** [comment]
@T-vK Could you add `-d` flag to `get-orig-source` so that it becomes: `debian/rules -d get-orig-source`

**Comment by T-vK** [comment]
@Eloston Thanks, that helped a lot! I managed to get rid of the error. 

This one: `E: ungoogled-chromium-browser source: uploader-address-malformed Maintainer <maintainer@null>` is gone now btw.
I'm getting this one though: `E: ungoogled-chromium-browser changes: changed-by-address-malformed ungoogled-chromium Authors <maintainer@null>`
https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/95187525

By replacing the `ungoogled-chromium Authors <maintainer@null>` from `packaging/ubuntu_bionic/changelog.ungoogin` with the the GPG key's user ID, I managed to get rid of that one as well.
https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/95194118

The file `ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc` seems to be missing:

```
Now signing changes and any dsc files...
 signfile dsc ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc tavk tavk <t-v.k@gmx.net>
gpg: signing failed: No such file or directory
gpg: /tmp/debsign.hwokAatw/ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc: clear-sign failed: No such file or directory
debsign: gpg error occurred!  Aborting....
debuild: fatal error at line 1081:
running debsign failed
```
  
And I still get these `source-is-missing` errors for a lot of files in the `third_party` directory.

I'll try to do an `ls` on a few of these pathes before the `debuild` to check if the files are actually missing.

**Comment by Eloston** [comment]
> The file ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc seems to be missing:

Try adding your signing identity to the `debian/control` file as well.

> edit: I couldn't find the location of the third_party directory. Do you know where it is?

There are multiple, but the most commonly used one is the one at the root of the Chromium source tree.

**Comment by T-vK** [comment]
@Eloston I also changed the identity in debian/control now, but I till get the error
https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/95610619

**Comment by Eloston** [comment]
Could you try replacing `debuild -S -sa` with `debuild -i -S`?

**Comment by T-vK** [comment]
Of course. 
I'm still getting the same error: https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/95980083

**Comment by Eloston** [comment]
Well, I'm out of ideas now. If I figure something out, I'll let you know.

**Comment by T-vK** [comment]
In the meantime I ran `find / -name "google-analytics-bundle.js"` before the `debuild` and also after the `debuild` fails and bot times I get no result. Not sure what to make out of that. I mean the error message does say `source-is-missing` , but at the same time it's talking about the line length of that file:

```
E: ungoogled-chromium-browser source: source-is-missing third_party/analytics/google-analytics-bundle.js line length is 525 characters (>512)
```

**Comment by Eloston** [comment]
> I mean the error message does say source-is-missing , but at the same time it's talking about the line length of that file

The file isn't removed by ungoogled-chromium commands, and I don't see any reference to it or its parent directory (`third_party/analytics`) within the `debian` directory either. It is also present in a fresh unpacking of the Chromium source tree.

However, it is domain substituted, which may cause the minified file to have lines that exceed 512 characters. But, I think it's more likely that it's normal to see these particular warnings since Chromium has minified files scattered about in the source code.

**Comment by T-vK** [comment]
Weird, I guess it's only temporarily being unpacked during the `debuild` command then? Anyway, I extracted the source tree and found the file in there. The line in question doesn't have any sort of domain string in it:
`(function() { 'use strict';var h,aa=aa||{},k=this,m=function(a){return void 0!==a},ba=function(){},ca=function(a){var b=typeof a;if("object"==b)if(a){if(a instanceof Array)return"array";if(a instanceof Object)return b;var c=Object.prototype.toString.call(a);if("[object Window]"==c)return"object";if("[object Array]"==c||"number"==typeof a.length&&"undefined"!=typeof a.splice&&"undefined"!=typeof a.propertyIsEnumerable&&!a.propertyIsEnumerable("splice"))return"array";if("[object Function]"==c||"undefined"!=typeof a.call&&`

The other files for which `source-is-missing` was reported seem to be there as well.

I also checked if the ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc is actually missing and it's not. It's located in `build/ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc` but `debuild` seems to expect it at `/tmp/debsign.YogMWoVr/ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc`. My guess is that it tries to copy the file to that tmp location and then sign it in there and finally copy it back. And the first copy step probably already fails. But I don't know...

**Comment by metsatron** [comment]
What is the current status of the PPA?

**Comment by T-vK** [comment]
Still stuck at that `gpg: /tmp/debsign.ndBgYzyR/ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc: clear-sign failed: No such file or directory` error. :/

**Comment by pcraciunoiu** [comment]
Subscribed, and

:+1: for anyone who can figure this out or provide instructions for figuring it quickly / others to help unblock

**Comment by podshumok** [comment]
@T-vK  you may want to add some information-gathering script (like `(pwd; lstree /) > lstree_gitliab_artifact` to debuild's `signing-hook` (see [man](https://manpages.debian.org/jessie/devscripts/debuild.1.en.html#HOOKS))

**Comment by T-vK** [comment]
I just tried. Unfortunately it didn't seem to work. 
I basically did this:
```
# export DEBUILD_SIGNING_HOOK='(pwd; lstree /) > lstree_gitliab_artifact.txt'
# debuild -i -S || echo
```
And then I checked for the file:
```
# ls -la
total 32
drwxr-xr-x. 4 root root 4096 Nov 18 06:37 .
drwxr-xr-x. 3 root root 4096 Nov 18 06:38 ..
drwxr-xr-x. 6 root root 4096 Nov 18 06:38 debian
drwxr-xr-x. 3 root root 4096 Nov 18 06:37 third_party
# cat lstree_gitliab_artifact.txt
cat: lstree_gitliab_artifact.txt: No such file or directory
```

**Comment by podshumok** [comment]
@T-vK 

hmm. I think it should be a separate script with +x attr and variable should have its path as a value

**Comment by T-vK** [comment]
@podshumok that doesn't appear to work either..
```
$ echo "#!/usr/bin/env bash" > signing-hook-script.sh
$ echo "(pwd; ls -la; tree /) > lstree_gitliab_artifact.txt" >> signing-hook-script.sh
$ chmod +x signing-hook-script.sh
$ export DEBUILD_SIGNING_HOOK='signing-hook-script.sh'
```
https://gitlab.com/T-vK/ungoogled-chromium-launchpad/-/jobs/122212458

Maybe the current directory is not the one I expected during the execution of the signing hook. I'll try to specify absolute paths.

**Comment by podshumok** [comment]
Well.
First, looks like I was wrong about environment variables.

HOOK should be specified in config file or as an argument to `debuild` script: `. For some reason documentation mentions this hook as `--signing-hook-foo` while I thing it should be `--signing-hook=/signing-hook-script.sh`

Second, looks like we need to see what `gpg` is doing:

```sh
gpg --verbose  --local-user "tavk tavk <t-v.k@gmx.net>" --clearsign \
    --list-options no-show-policy-urls \
    --armor --textmode --output "/build/test-signature"\
    "/build/ungoogled-chromium-browser_*.dsc"
```
(or maybe try to sign any other file)

**Comment by T-vK** [comment]
@podshumok 
The man page is indeed wrong. 
Here is the output of the signing hook. (`ls -la /` followed by `tree /`) https://t-vk.gitlab.io/-/ungoogled-chromium-launchpad/-/jobs/123825062/artifacts/build/lstree_gitliab_artifact.txt

While attempting to sign a file using gpg I made a very helpful observation.
Then the file to be signed does not exist you get an error like this:
```
gpg: can't open '/non-existing-file.txt': No such file or directory
gpg: /non-existing-file.txt: clear-sign failed: No such file or directory
```

And when only the output directly does not exist, you get an error like this:
```
gpg: writing to '/non-existing-output-dir'
gpg: pinentry launched (585 curses 1.1.0 not a tty - -)
gpg: signing failed: No such file or directory
gpg: /existing-file.txt: clear-sign failed: No such file or directory
```

Comparing that to the error that is currently preventing the ungoogled chromium file from being signed: 
```
gpg: signing failed: No such file or directory
gpg: /tmp/debsign.hwokAatw/ungoogled-chromium-browser_68.0.3440.106-1~bionic.dsc: clear-sign failed: No such file or directory
```

it becomes clear that the issue is most likely not that the .dsc file is missing, but that the output directory that debuild passes to gpg doesn't exist.

Anyway,

**Comment by T-vK** [comment]
The devscripts package version is `2.17.12ubuntu1.1`.  
Unfortunately `mktemp` seems to work :(
```
$ TMP_FILE="$(mktemp)"
$ ls -la /tmp
total 20
drwxrwxrwt. 1 root root 4096 Nov 23 10:17 .
drwxr-xr-x. 1 root root 4096 Nov 23 10:17 ..
-rw-------. 1 root root    0 Nov 23 10:17 tmp.nKYuLBJlJm
$ echo "test 123" >> "$TMP_FILE"
$ ls -la "$TMP_FILE"
-rw-------. 1 root root 9 Nov 23 10:17 /tmp/tmp.nKYuLBJlJm
$ cat "$TMP_FILE"
test 123
```

**Comment by T-vK** [comment]
@Eloston Hey, I just wanted to give it another try with the current master and I see a lot of things have changed. The CI script used to replace some values in `packaging/ubuntu_bionic/changelog.ungoogin`, so that they match the GPG key, but that file does not exist anymore. Is there a different file that takes care of this now?

**Comment by Eloston** [comment]
@T-vK All Debian code has moved to [ungoogled-chromium-debian](https://github.com/ungoogled-software/ungoogled-chromium-debian). The file structure follows that of Debian's `chromium` package git repo.

FYI, I don't really mind if we keep this issue open or open a new one on that repo. Unfortuantely, GitHub can't transfer this issue over to ungoogled-chromium-debian.

**Comment by podshumok** [comment]
Hi! Here is how vscodium guys solved this puzzle https://gitlab.com/paulcarroty/vscodium-deb-rpm-repo/blob/master/updaterepos.sh

**Comment by Eloston** [comment]
@podshumok That's quite the hack they've put together, but it's not very robust. [At least on GitHub, there is a git blob limit of 100 MB](https://help.github.com/en/articles/what-is-my-disk-quota) (we aren't hitting it *yet*, though). Also, their solution involves uploading the **private** .deb signing key to the repo, which is less than an ideal solution for security reasons (users would need to trust the public key, which opens them up to security problems). If there's a good solution to the private key security problem, then we can adopt their solution here.

**Comment by podshumok** [comment]
The key is encrypted and they decrypt it using CI secrets. I think it is more or less fine, but it also should be possible to have private key completely stored in CI secrets. What I don't like in their solution is the way original deb files are downloaded: they just `curl` github releases page and `grep` links to the latest deb's. I think at least some sort of verification is required here - maybe signing. In that case if for example their github account gets hacked authors still can revoke the key and my system wouldn't get forged package update.

**Comment by Eloston** [comment]
@podshumok 

> The key is encrypted and they decrypt it using CI secrets

Ah, I missed that crucial detail in my haste. Thanks. In that case, that part is fine as long as there's strong security around how those variables can be used; e.g. if the variables can be used in PRs, then maybe it's possible modify the CI script in the PR to redirect the key somewhere else.

> What I don't like in their solution is the way original deb files are downloaded: they just curl github releases page and grep links to the latest deb's. I think at least some sort of verification is required here - maybe signing

At that point, it's hard to say if the repo containing the key to decrypt the private key is any safer. In our scenario, we are trusting that all those pushing binaries to ungoogled-chromium-binaries will keep their accounts secure. If we really want a secure pipeline for building and publishing binaries, we should consider automating the build on a more trustworthy server.

However, using an encrypted private key and retrieving deb packages from ungoogled-chromium-binaries as it is now is not much more dangerous than users manually downloading the deb packages themselves. In addition, the largest deb packages we have now are around 50 MB, which is still well below the limit of 100 MB. For now, we could implement a solution like the one you found.

**Comment by mckaygerhard** [comment]
PPA are not good idea.. due only winbutu shit will be easy to setup.. all of the debian derivatives will not be easy installable for newbie Debian or realted Debian users.. 

so OBS will benefit all the required if a proper rules are made.. of course the build enviroment are too exquisite.. that's the problem.. (specific python version .. specific engines for realted libraries etc)

**Comment by Qik000** [comment]
A repo would be a very good idea, the updates woundnt require manual installation and the install itself would be eaaier

**Comment by trymeouteh** [comment]
I would like to see this app preferably in the debian repos and always be the latest version. But if there can be PPA that is the latest version then that is good enough for me.

I know I can get the latest version if I switched to Arch based distro but I would like the latest version of Ungoogled Chromium on my Debian/Ubuntu based distro since Debian and Ubuntu based distro are more stable compared to Arch and I like using a stable distro for my work computer.

**Comment by hrj** [comment]
Replying to @mckaygerhard comment:
> of course the build enviroment are too exquisite.. that's the problem.. (specific python version .. specific engines for realted libraries etc)

Solution could be to use a container based build. The container has all the dependencies, and doesn't depend on or pollute the host. I made a start here: https://github.com/ungoogled-software/ungoogled-chromium-portablelinux/pull/22/files

The build just requires a single command to be executed, and a single dependency on docker. Since this is the portable version, the compiled binary will work on all linux platforms, and since the build is inside a container, it will build on all linux platforms.

If we all focus on this path, we could reduce the effort involved in building, as people using different distributions could use the same build scripts.

Eventually, we can derive debs and other packages out of these binaries. These may not be acceptable to official repositories (because of bundled libraries) but might help us get there eventually, by getting more adoption.

**Comment by markstos** [comment]
There is a Ubuntu PPA for this now but oddly this is the first result when you search for it and no one here seems to mentioned the link. It's here:

https://launchpad.net/~braewoods/+archive/ubuntu/ungoogled-chromium

**Comment by mbakke** [comment]
An alternative way to get the latest ungoogled-chromium on any GNU/Linux distribution is to [install GNU Guix](https://guix.gnu.org/manual/en/guix.html#Binary-Installation).  Then you can simply `guix install ungoogled-chromium` (and any other Guix package), no root privileges needed.

**Comment by trymeouteh** [comment]
> There is a Ubuntu PPA for this now but oddly this is the first result when you search for it and no one here seems to mentioned the link. It's here:
> 
> https://launchpad.net/~braewoods/+archive/ubuntu/ungoogled-chromium

Is there a PPA without any added extensions like keepassxc? I want a plain vanilla ungoogled chromium PPA that keeps ungoogled chromium up to date, no added extensions or anything.

**Comment by braewoods** [comment]
@trymeouteh why is that a problem? They're not pulled in when you install the main packages. They're only intended for people that would like a systematic replacement for automatic extension updates which appears to not work otherwise with ungoogled-chromium.

**Comment by nicoroeser** [comment]
@braewoods, has your PPA been moved or removed?

**Comment by braewoods** [comment]
@nicoroeser It was replaced by the OBS repository that I now manage via automated uploads from the official github Debian repository. Switch over to using that and you should be able to update just fine. Just note that it doesn't provide extensions anymore as I found out you apparently don't need that. There's better ways to install them from what I was reading.

**Comment by Eloston** [comment]
Closing because OBS is rolled out.
